tensorpack/train/trainers.py

```diff
@@ -403,23 +416,88 @@ def allreduce(self, grads):
         with tf.name_scope("AllReduce"):
             for grad, var in grads:
                 if grad is not None:
+                    scaled_grad = tf.math.divide(grad, self._aggregation_frequency)
                     if self._compression is not None and self._has_compression:
-                        avg_grad = self.hvd.allreduce(grad, average=self._average, compression=self._compression)
+                        avg_grad = self.hvd.allreduce(scaled_grad, average=self._average, compression=self._compression)
                     else:
-                        avg_grad = self.hvd.allreduce(grad, average=self._average)
+                        avg_grad = self.hvd.allreduce(scaled_grad, average=self._average)
                     averaged_gradients.append((avg_grad, var))
                 else:
                     averaged_gradients.append((None, var))
         return averaged_gradients
 
     def _setup_graph(self, input, get_cost_fn, get_opt_fn):
         with TrainTowerContext(''):
-            grads = self._make_get_grad_fn(input, get_cost_fn, get_opt_fn)()
-            grads = self.allreduce(grads)
-
-            opt = get_opt_fn()
-            self.train_op = opt.apply_gradients(grads, name='train_op')
-            self.comm_op = tf.no_op()
+            self.grads = self._make_get_grad_fn(input, get_cost_fn, get_opt_fn)()
+            aggregation_ops_list = []
+            if self._aggregation_frequency > 1:
+                # Create the variables which will be used to store aggregated updates.
+                with tf.variable_scope("aggregation_variables"):
+                    for idx, (grad, var) in enumerate(self.grads):
+                        grad_aggregation_variable_name = str(idx)
```

Comment:
Could be useful for a comment here about why it's ok to use the idx for the name versus what we had done previously.